add documentation to builder pattern (#415)
Codecademy/docs: content/general/concepts/what-is-the-builder-pattern/what-is-the-builder-pattern.md

---
Title: 'What is the Builder Pattern?'
Description: 'The builder pattern decouples the responsibility of object creation from the desired objects class.'

> Subjects:
> - 'Computer Science'
> - 'Interview Prep' Tags:
> - 'Classes'
> - 'Conceptual'
> - 'Constructors'
> - 'Objects' CatalogContent:
> - 'learn-java'
> - 'paths/computer-science'

Reviewer: ```suggestion
Subjects:
  - 'Computer Science'
  - 'Interview Prep' 
Tags:
  - 'Classes'
  - 'Conceptual'
  - 'Constructors'
  - 'Objects'
CatalogContent:
  - 'learn-java'
  - 'paths/computer-science'
```

> ---

Reviewer: ```suggestion
---

```

> The _builder pattern_ decouples the responsibility of object creation from the desired object's class. It is normally implemented when a class requires many properties to be set during it's construction and may be used to ensure immutability. 

Reviewer: ```suggestion
The _builder pattern_ decouples the responsibility of object creation from the desired object's class. It is normally implemented when a class requires many properties to be set during its construction and may be used to ensure immutability. 
```

## UML Design

![UML diagram of a builder](media/builder-uml.png)

## Example

> To illustrate the _builder pattern_, below provides a real world example, in Java, depicting potential considerations for a pizza order. The PizzaOrder class has 5 properties; toppings, cheese, sauce, pizzaBase and stuffCrust. During construction, it requires both pizzaBase and stuffCrust to be set explicitly, and should cheese and / or sauce not be set before build(), default values will be assigned. Construction of a PizzaOrder object is enforced through the PizzaOrder's Builder inner class. After creation, a PizzaOrder object's properties are immutable.

Reviewer: ```suggestion
To illustrate the _builder pattern_, below provides a real-world example in Java, depicting potential considerations for a pizza order. The `PizzaOrder` class has 5 properties: `toppings`, `cheese`, `sauce`, `pizzaBase` and `stuffCrust`. During construction, both `pizzaBase` and `stuffCrust` are required to be set explicitly, and should cheese and/or sauce not be set before `build()`, default values will be assigned. Construction of a `PizzaOrder` object is enforced through the `PizzaOrder`'s `Builder` inner class. After creation, a `PizzaOrder` object's properties are immutable.
```

            return this;
        }
        
        public Builder sauce(Sauce sauce) { 
            this.sauce = sauce;
            return this;
        }
        
        public Builder base(Base pizzaBase) { 
            this.pizzaBase = pizzaBase;
            return this;
        }
        
        public Builder stuffCrust(boolean stuffCrust) { 
            this.stuffCrust = stuffCrust;
            return this;
        }

        // When build() is called, a new concrete object is returned with the desired properties set
        public PizzaOrder build() { return new PizzaOrder(this); }
    }
    
    private final List<Topping> toppings;
    private final Cheese cheese;
    private final Sauce sauce;
    private final Base pizzaBase;
    private final boolean stuffCrust;
    
    // Removes default constructor and forces object creation through Builder inner class
    private PizzaOrder(Builder builder) {
        // Constructs concrete object with builder's property values
        this.toppings = builder.toppings;
        this.cheese = builder.cheese;
        this.sauce = builder.sauce;
        this.stuffCrust = builder.stuffCrust;
    }
    

>     // Getters below. (Note: setters negated for immutability.)

Reviewer: We can also place notes throughout an entry like so (not within code snippets):

**Note:** The contents of your note here.

}
```

> Note: Some Enums have been provided to safely model the choices an order may have. 

Reviewer: ```suggestion
**Note:** Some Enums have been provided to safely model the choices an order may have. 
```

```java
public enum Topping
{

>     HAM,
>     PINEAPPLE,
>     SWEETCORN

Reviewer: ```suggestion
  HAM,
  PINEAPPLE,
  SWEETCORN
```

Style: 2 spaces rather than 1 tab/ 4 spaces

}
```

```java
public enum Sauce {

>     TOMATO,
>     BBQ

Reviewer: ```suggestion
  TOMATO,
  BBQ
```

}
```

```java
public enum Base {

>     THIN_BASE,
>     DEEP_PAN

Reviewer: ```suggestion
  THIN_BASE,
  DEEP_PAN
```

}
```

```java
public enum Cheese {

>     NONE,
>     MOZZARELLA

Reviewer: ```suggestion
  NONE,
  MOZZARELLA
```

}
```

Note: The Main class below shows how to instanciate a PizzaOrder object through its inner Builder class.

```java
public class Main {

>     public static void main(String[] args) {
>         // Create concrete object
>         PizzaOrder pizzaOrder = new PizzaOrder.Builder(Base.DEEP_PAN, true).sauce(Sauce.BBQ)
>                                                                            .build();
>     }

Reviewer: ```suggestion
  public static void main(String[] args) {
    // Create concrete object
    PizzaOrder pizzaOrder = new PizzaOrder.Builder(Base.DEEP_PAN, true).sauce(Sauce.BBQ)
                                                                           .build();
  }
```

Description: 'The builder pattern decouples the responsibility of object creation from the desired objects class.'
Subjects:
- 'Computer Science'
- 'Interview Prep' Tags:
- 'Classes'
- 'Conceptual'
- 'Constructors'
- 'Objects' CatalogContent:
- 'learn-java'
- 'paths/computer-science'
---
The _builder pattern_ decouples the responsibility of object creation from the desired object's class. It is normally implemented when a class 
requires many properties to be set during it's construction and may be used to ensure immutability. 

## UML Design

![UML diagram of a builder](media/builder-uml.png)

## Example

To illustrate the _builder pattern_, below provides a real world example, in Java, depicting potential considerations for a pizza order. The 
PizzaOrder class has 5 properties; toppings, cheese, sauce, pizzaBase and stuffCrust. During construction, it requires both pizzaBase and stuffCrust 
to be set explicitly, and should cheese and / or sauce not be set before build(), default values will be assigned. Construction of a PizzaOrder 
object is enforced through the PizzaOrder's Builder inner class. After creation, a PizzaOrder object's properties are immutable.

> ```java

Reviewer 2: ```suggestion

```java
```

Author: Completed in other comment

        public Builder toppings(List<Topping> toppings) {
            this.toppings = toppings;
            return this;
        }
        
        public Builder cheese(Cheese cheese) { ... }
        
        public Builder sauce(Sauce sauce) { ... }
        
        public Builder base(Base pizzaBase) { ... }
        
        public Builder stuffCrust(boolean stuffCrust) { ... }

        // When build() is called, a new concrete object is returned with the desired properties set
        public PizzaOrder build() { return new PizzaOrder(this); }
    }
    
    private final List<Topping> toppings;
    private final Cheese cheese;
    private final Sauce sauce;
    private final Base pizzaBase;
    private final boolean stuffCrust;
    
    // Removes default constructor and forces object creation through Builder inner class
    private PizzaOrder(Builder builder) {
        // Constructs concrete object with builder's property values
        this.toppings = builder.toppings;
        this.cheese = builder.cheese;
        this.sauce = builder.sauce;
        this.stuffCrust = builder.stuffCrust;
    }
    
    // Getters below. (Note: setters negated for immutability.)
}
```

> ```java

Reviewer 2: ```suggestion

```java
```

---
Title: 'What is the Builder Pattern?'
Description: 'The builder pattern decouples the responsibility of object creation from the desired objects class.'
Subjects:
- 'Computer Science'
- 'Interview Prep'

>   Tags:

Reviewer 2: ```suggestion
Tags:
```

- 'Classes'
- 'Conceptual'
- 'Constructors'
- 'Objects'

>   CatalogContent:

Reviewer 2: ```suggestion
CatalogContent:
```

- 'learn-java'
- 'paths/computer-science'
---

> The _builder pattern_ decouples the responsibility of object creation from the desired object's class. It is normally implemented when a class 

Reviewer 2: ```suggestion
The _builder pattern_ decouples the responsibility of object creation from the desired object's class. It is normally implemented when a class requires many properties to be set during it's construction and may be used to ensure immutability. 
```

> requires many properties to be set during it's construction and may be used to ensure immutability. 

Reviewer 2: ```suggestion
```

## UML Design

> ![UML diagram of a builder](media/builder-uml.png)

Reviewer 2: Check this file. What is uploaded doesn't seem right. Just two headers and blank space.

Author: It looks okay on my end. The only thing I can think of is the two headers are filled in white where as the rest of the info is written in black on a transparent background. Dark mode may make this difficult to see. Please see below:

![image](https://user-images.githubusercontent.com/78788406/140649298-36b96315-fa9b-4353-9185-7222a4f8fba7.png)

Please advise?

Reviewer 2: Can you add a white background?

## Example

> To illustrate the _builder pattern_, below provides a real world example depicting potential considerations for a pizza order. The PizzaOrder class 

Reviewer 2: ```suggestion
To illustrate the _builder pattern_, below provides a real world example, in Java, depicting potential considerations for a pizza order. The PizzaOrder class 
```

> has 5 properties; toppings, cheese, sauce, pizzaBase and stuffCrust. During construction, we require both pizzaBase and stuffCrust to be set

Reviewer 2: ```suggestion
```

> explicitly, and should cheese and / or sauce not be set before build(), default values will be assigned. Construction of a PizzaOrder object is 

Reviewer 2: ```suggestion
```

> enforced through the PizzaOrder's Builder inner class, and post creation, a PizzaOrder object's properties are immutable.

Reviewer 2: ```suggestion
```

Author: Hi,

I'm getting slightly confused with these comments on the description of the example. Is the issue the amount of characters I have in a line?

Thanks

Reviewer 2: My preference is generally to have paragraphs on a single line.

```java
public class PizzaOrder {
    public static class Builder {
        // Same properties as the outer PizzaOrder class
        private final List<Topping> toppings;
        private final Cheese cheese;
        private final Sauce sauce;
        private final Base pizzaBase;
        private final boolean stuffCrust;
        
        public Builder(Base pizzaBase, boolean stuffCrust) {
            // Enforce required choices
            this.pizzaBase = pizzaBase;
            this.stuffCrust = stuffCrust;
            // Default values
            this.cheese = Cheese.MOZZARELLA;

>             this.sauce = Sauce.TOMATO;

Reviewer 2: Should define the enums in the example as well, so the code works as listed. The enums don't have to be exhaustive, just enough that the example works and to give an idea how the example works.

        }
        
        // Sets property and returns itself (either for next property to be set, or build() to be invoked)
        public Builder toppings(List<Topping> toppings) {
            this.toppings = toppings;
            return this;
        }
        

>         public Builder cheese(Cheese cheese) { ... }

Reviewer 2: Shouldn't omit code in what should be a working example.

        
        public Builder sauce(Sauce sauce) { ... }
        
        public Builder base(Base pizzaBase) { ... }
        
        public Builder stuffCrust(boolean stuffCrust) { ... }

        // When build() is called, a new concrete object is returned with the desired properties set
        public PizzaOrder build() { return new PizzaOrder(this); }
    }
    
    private final List<Topping> toppings;
    private final Cheese cheese;
    private final Sauce sauce;
    private final Base pizzaBase;
    private final boolean stuffCrust;
    
    // Removes default constructor and forces object creation through Builder inner class
    private PizzaOrder(Builder builder) {
        // Constructs concrete object with builder's property values
        this.toppings = builder.toppings;
        this.cheese = builder.cheese;
        this.sauce = builder.sauce;
        this.stuffCrust = builder.stuffCrust;
    }
    
    // Getters below. (Note: setters negated for immutability.)
}
```
```java
public class Main {
    public static void main(String[] args) {
        // Create concrete object

>         PizzaOrder pizzaOrder = new PizzaOrder.Builder(Base.DEEP_PAN, true).sauce(sauce.BBQ)

Reviewer 2: Should define the enums in the example as well, so the code works as listed.  The enums don't have to be exhaustive, just enough that the example works and to give an idea how the example works.
